src/backend/distributed/catalog/dependency.c

```diff
@@ -0,0 +1,381 @@
+ *  - object is not created by an other extension. Objects created by extensions are
+ *    assumed to be created on the worker when the extension is created there.
+ *  - citus supports object distribution. If citus does not support the distribution of
+ *    the object we will not try and distribute it.
+ */
+static bool
+follow_get_dependencies_for_object(void *context, const Form_pg_depend pg_depend)
+{
+	List **targetList = (List **) context;
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, pg_depend->refclassid, pg_depend->refobjid);
+
+	if (pg_depend->deptype != DEPENDENCY_NORMAL)
+	{
+		return false;
+	}
+
+	if (IsObjectAddressInList(&address, *targetList))
+	{
+		return false;
+	}
+
+	if (isObjectDistributed(&address))
+	{
+		return false;
+	}
+
+	if (IsObjectAddressOwnedByExtension(&address))
+	{
+		return false;
+	}
+
+	if (!SupportedDependencyByCitus(&address))
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * follow_order_object_address applies filters on pg_depend entries to follow the
+ * dependency tree of objects in depth first order. The filters are practically the same
+ * to follow_get_dependencies_for_object, except it will follow objects that have already
+ * been distributed. This is because its primary use is to order the objects from
+ * pg_dist_object in dependency order.
+ *
+ * For completeness the list of all edges it will follow;
+ *
+ * Objects are only followed if all of the following checks hold true:
+ *  - the pg_depend entry is a normal dependency, all other types are created and
+ *    maintained by postgres
+ *  - the object is not already in the targetList (context), if it is already in there an
+ *    other object already caused the creation of this object
+ *  - object is not created by an other extension. Objects created by extensions are
+ *    assumed to be created on the worker when the extension is created there.
+ *  - citus supports object distribution. If citus does not support the distribution of
+ *    the object we will not try and distribute it.
+ */
+static bool
+follow_order_object_address(void *context, const Form_pg_depend pg_depend)
+{
+	List **targetList = (List **) context;
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, pg_depend->refclassid, pg_depend->refobjid);
+
+	if (pg_depend->deptype != DEPENDENCY_NORMAL)
+	{
+		return false;
+	}
+
+	if (IsObjectAddressInList(&address, *targetList))
+	{
+		return false;
+	}
+
+	if (IsObjectAddressOwnedByExtension(&address))
+	{
+		return false;
+	}
+
+	if (!SupportedDependencyByCitus(&address))
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * apply_add_to_target_list is an apply function for recurse_pg_depend that will append
+ * all the ObjectAddresses for pg_depend entries to the context. The context here is
+ * assumed to be a (List **) to the location where all ObjectAddresses will be collected.
+ */
+static void
+apply_add_to_target_list(void *context, const Form_pg_depend pg_depend)
```

Comment:
The term `TargetList` is used in postgres' `queryTree->targetList`, and it is one of the common parts of the queryTree that Citus interacts with. So, it is very much used in other places.

Can we use another name here such as `dependencyList` or `requiredDependcyList` etc? I think if you change it, you'd need to go through the code and update some other places that use targetList.

Reply:
Good point, forgot that target list is pretty common in postgres. Will go with a simple dependencyList.